warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -32,174 +33,303 @@
+      if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
+        if (null != splitTypeLabel) {
+          split.getMetadata().putToLabels(splitTypeLabel, SPLIT_TYPE_DISTANCE);
+        }
         mustSplit = true;
       }
-      
+
       //
-      // If the current point is farther away from the reference point than 'proximityZoneRadius'
+      // If the current point is farther away from the reference point than 'proximityZoneRadius', or the instant speed is greater than proximityZoneSpeed
       // change the reference point and reset the traveled distance.
-      // If the distance traveled in the proximity zone was traveled at less than proximityZoneSpeed and the time spent in the zone is above proximityZoneTime, force a split
+      // If the distance traveled in the proximity zone was traveled at less than proximityInZoneMaxMeanSpeed and the time spent in the zone is above proximityZoneTime, force a split
       //
-      
-      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) > proximityZoneRadius) {
-        if (previousTimestamp - refTimestamp > proximityZoneTime && proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S) < proximityZoneSpeed) {
-          mustSplit = true;
+      long timeStopped = previousTimestamp - refTimestamp;
+      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && null != split) {
+
+        double currentSpeed = 0.0D;
+        if (GeoTimeSerie.NO_LOCATION != previousLocation && timestamp != previousTimestamp) {
+          currentSpeed = GeoXPLib.orthodromicDistance(previousLocation, location) / ((double) (timestamp - previousTimestamp) / Constants.TIME_UNITS_PER_S);
         }
-        refLocation = location;
-        refTimestamp = timestamp;
-        proximityZoneTraveledDistance = 0.0D;
-      } else if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) <= proximityZoneRadius) {
-        if (GeoTimeSerie.NO_LOCATION != previousLocation) {
-          proximityZoneTraveledDistance += GeoXPLib.orthodromicDistance(previousLocation, location);
+
+        // quit the radius, or speed greater than min speed.
+        if (GeoXPLib.orthodromicDistance(refLocation, location) > proximityZoneRadius || currentSpeed > proximityZoneMaxSpeed) {
+          double zoneMeanSpeed = 0.0D;
+          if (previousTimestamp != refTimestamp) {
+            zoneMeanSpeed = proximityZoneTraveledDistance / ((double) (previousTimestamp - refTimestamp) / Constants.TIME_UNITS_PER_S);
+          }
+          if (timeStopped > proximityZoneTime && zoneMeanSpeed < proximityInZoneMaxMeanSpeed) {
```

Comment:
Actually this should be checked even if still in the proximity zone

Reply:
High deceleration case: stops in less than radius : On the next iteration, the mean speed will be high too. That's why I added a separate max stopped speed, which is always checked.
What could be done next is to trig on the derivate (accerelation)